op-supervisor/supervisor/backend/rewinder/rewinder.go

```diff
@@ -0,0 +1,232 @@
+package rewinder
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/ethereum/go-ethereum/log"
+
+	"github.com/ethereum-optimism/optimism/op-node/rollup/event"
+	"github.com/ethereum-optimism/optimism/op-service/eth"
+	"github.com/ethereum-optimism/optimism/op-service/locks"
+	"github.com/ethereum-optimism/optimism/op-supervisor/supervisor/backend/superevents"
+	"github.com/ethereum-optimism/optimism/op-supervisor/supervisor/types"
+)
+
+type rewinderDB interface {
+	LocalUnsafe(eth.ChainID) (types.BlockSeal, error)
+	CrossUnsafe(eth.ChainID) (types.BlockSeal, error)
+	LocalSafe(eth.ChainID) (types.DerivedBlockSealPair, error)
+	CrossSafe(eth.ChainID) (types.DerivedBlockSealPair, error)
+
+	RewindLocalUnsafe(eth.ChainID, types.BlockSeal) error
+	RewindCrossUnsafe(eth.ChainID, types.BlockSeal) error
+	RewindLocalSafe(eth.ChainID, types.BlockSeal) error
+	RewindCrossSafe(eth.ChainID, types.BlockSeal) error
+
+	FindSealedBlock(eth.ChainID, uint64) (types.BlockSeal, error)
+	Finalized(eth.ChainID) (types.BlockSeal, error)
+}
+
+type syncNode interface {
+	BlockRefByNumber(ctx context.Context, number uint64) (eth.BlockRef, error)
+}
+
+// rewindController holds the methods for a rewind operation
+type rewindController struct {
+	isSafe      bool
+	getLocal    func(eth.ChainID) (types.BlockSeal, error)
+	getCross    func(eth.ChainID) (types.BlockSeal, error)
+	rewindLocal func(eth.ChainID, types.BlockSeal) error
+	rewindCross func(eth.ChainID, types.BlockSeal) error
+}
+
+func (rc rewindController) safetyStr() string {
+	if rc.isSafe {
+		return "safe"
+	}
+	return "unsafe"
+}
+
+// Rewinder is responsible for handling the rewinding of databases to the latest common ancestor between
+// the local databases and L2 node.
+type Rewinder struct {
+	log       log.Logger
+	emitter   event.Emitter
+	db        rewinderDB
+	syncNodes locks.RWMap[eth.ChainID, syncNode]
+}
+
+func New(log log.Logger, db rewinderDB) *Rewinder {
+	return &Rewinder{
+		log: log.New("component", "rewinder"),
+		db:  db,
+	}
+}
+
+func (r *Rewinder) AttachEmitter(em event.Emitter) {
+	r.emitter = em
+}
+
+func (r *Rewinder) OnEvent(ev event.Event) bool {
+	switch x := ev.(type) {
+	case superevents.RewindChainEvent:
+		r.handleRewindChainEvent(x)
+		return true
+	case superevents.RewindAllChainsEvent:
+		r.handleRewindAllChainsEvent(x)
+		return true
+	default:
+		return false
+	}
+}
+
+func (r *Rewinder) AttachSyncNode(chainID eth.ChainID, source syncNode) {
+	r.syncNodes.Set(chainID, source)
+}
+
+func (r *Rewinder) handleRewindChainEvent(ev superevents.RewindChainEvent) {
+	if err := r.rewindChain(ev); err != nil {
+		r.log.Error("failed to rewind chain %s: %w", ev.ChainID, err)
+	}
+}
+func (r *Rewinder) handleRewindAllChainsEvent(ev superevents.RewindAllChainsEvent) {
+	r.syncNodes.Range(func(chainID eth.ChainID, source syncNode) bool {
+		r.emitter.Emit(superevents.RewindChainEvent{
+			ChainID:        chainID,
+			BadBlockHeight: ev.BadBlock.Number,
```

Comment:
This is wrong; I'll fix. We're sending an L1 block ref so we should be finding the corresponding L2 block to use here instead.